### ISSUE TITLE
Remove the `Magento\Framework\` entry from the psr4 autoload section

### ISSUE THREAD

**Issue by thlassche** [magento/magento2#34031]
### Summary (*)
By default, Magento's composer.json file contains a lib/Framework autoloader line: https://github.com/magento/magento2/blob/2.4-develop/composer.json#L355

This makes autoloading every framework file with a non-optimized Composer autoloader 50% slower, as first the file is checked for existince in`lib/internal/Magento/Framework/` and then in `vendor/magento/framework`. This leads to two `file_exists` calls for every framework file.

In my case, there is not a single PHP file in `lib/internal/Magento/Framework` (as is the case with default meta package install mentioned over here: https://devdocs.magento.com/guides/v2.4/install-gde/composer.html), hence the line makes no sense.

**Why not keep it there?**

When developing locally / in docker with Magento, performance usually is a real pain and slowing down development. This fix leads to 50% faster framework setup time for me. So if I e.g. clear `pub/static/frontend` (which I do every now and then while doing frontend development), I have to wait 50% less time

**What is the benefit of removing the line?**

- In my testing, e.g. generating a static file is 50% faster in M2 developer mode, see also the examples section

### Examples (*)
Blackfire results with and without the autoloading line removed:

<img width="561" alt="Screenshot 2021-09-09 at 11 48 03" src="https://user-images.githubusercontent.com/2959888/132664903-fc492434-0bb8-4d90-80d3-f4d30b6a1f71.png">

### Proposed solution
Remove `https://github.com/magento/magento2/blob/2.4-develop/composer.json#L355`

### Additional Information

Removing the line `lib/internal/Magento/Framework/` from here:

https://github.com/magento/magento2/blob/9aa35fd7df8e0a42e5a2a32c8ad3c5b895346091/composer.json#L365

And also as @hostep said removing the line `app/code/Magento/` from here:

https://github.com/magento/magento2/blob/9aa35fd7df8e0a42e5a2a32c8ad3c5b895346091/composer.json#L367

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [x] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by thlassche** [comment]
@hostep @ihor-sviziev @barryvdh @qsolutions-pl You guys might also find this interesting

**Comment by ihor-sviziev** [comment]
@thlassche, will it work with GitHub-based installation?

**Comment by thlassche** [comment]
> @thlassche, will it work with GitHub-based installation?

My suggestion would be to change it in the meta package only. I.e. the one you get when you run

```
composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition <name>
```

**Comment by hostep** [comment]
That sounds like a really good idea to me!

@fascinosum: is something like this possible in the packaging of the meta package?

**Comment by thlassche** [comment]
> That sounds like a really good idea to me!
> 
> @fascinosum: is something like this possible in the packaging of the meta package?
> 
> _Update_, we can then probably also remove this line as well:
> 
> ```
>             "Magento\\": "app/code/Magento/",
> ```
> 
> Since that directory is empty in the composer install of a Magento shop, every Magento module is located in `vendor`

Agreed!

**Comment by fascinosum** [comment]
hi @thlassche, @hostep, I think it makes sense to remove these 2 namespace declarations from the project metapackage
@hostep, it is possible, but it needs to be implemented :)
As everyone knows, this change, if implemented, will take effect only on new installations as a project. `magento/composer-root-update-plugin` should help apply this change for updated projects
Each project that uses only `magento/product-<edition>-edition` will not be affected (will not benefit) from this change
For example, Cloud template does not have one of the above declarations https://github.com/magento/magento-cloud/blob/master/composer.json
cc: @tpogrebniak @pdohogne-magento

**Comment by pdohogne-magento** [comment]
> Each project that uses only `magento/product-<edition>-edition` will not be affected (will not benefit) from this change. For example, Cloud template does not have one of the above declarations

One clarification: `magento/composer-root-update-plugin` does work for Cloud projects that require `magento/magento-cloud-metapackage` instead of `magento/product-<edition>-edition` (the plugin specifically looks for either a `magento/product-*` or cloud metapackage requirement, see [the documentation](https://github.com/magento/composer-root-update-plugin/blob/1.1/src/Magento/ComposerRootUpdatePlugin/README.md#purpose-of-plugin)).  So while the Cloud template does not contain `Magento\Framework\` in `psr-4` and won't need that particular change, the plugin would pick up the removal of `app/code/Magento` if that change is made in a new template release.

**Comment by Den4ik** [comment]
Interested optimization

**Comment by torhoehn** [comment]
> That sounds like a really good idea to me!
> 
> @fascinosum: is something like this possible in the packaging of the meta package?
> 
> _Update_, we can then probably also remove this line as well:
> 
> ```
>             "Magento\\": "app/code/Magento/",
> ```
> 
> Since that directory is empty in the composer install of a Magento shop, every Magento module is located in `vendor`

@hostep So you will force to implement own modules as composer requirement? Maybe this needs custom infrastructure on integrator side.

**Comment by hostep** [comment]
@torhoehn: this is only for modules that are using the Magento namespace. And all of those get installed via composer, right?

**Comment by torhoehn** [comment]
@hostep Ah sorry, you're absolutely right!

**Comment by ihor-sviziev** [comment]
@sidolov, I set priority to P1 because it supposes to have a significant performance improvement

**Comment by engcom-Hotel** [comment]
Hello @thlassche,

Thanks for the report and collaboration!

We have looked into the issue and found that if we made the said changes in the main description will improve the Magento performance during development. 

Removing the line `lib/internal/Magento/Framework/` from here:

https://github.com/magento/magento2/blob/9aa35fd7df8e0a42e5a2a32c8ad3c5b895346091/composer.json#L365

And also as @hostep said removing the line `app/code/Magento/` from here:

https://github.com/magento/magento2/blob/9aa35fd7df8e0a42e5a2a32c8ad3c5b895346091/composer.json#L367

will increase the performance. Because these paths are working for the Github installation only and not for Composer installation. All the files listed in the above folders are in the vendor folder during composer installation. 

Hence confirming the issue.

Thanks

**Comment by hostep** [comment]
Hi all

Are there plans to move forward with this? Anybody an idea what the status of internal ticket AC-6621 would be?

Probably the Adobe packaging team (@rjtulloh maybe?) will need to get involved here and since the packaging system isn't open sourced, it will be hard for the community to contribute to this ...

**Comment by hostep** [comment]
Looks like this will happen in Magento 2.4.7

Comparing 2.4.6-p4 with 2.4.7-beta3 installed via composer:
```diff
             ]
         },
         "psr-4": {
-            "Magento\\": "app/code/Magento/",
-            "Magento\\Framework\\": "lib/internal/Magento/Framework/",
             "Magento\\Setup\\": "setup/src/Magento/Setup/"
         }
     },
```

**Comment by hostep** [comment]
Confirming that this is also included in the final version of 2.4.7

So this issue can be closed now I guess? 🙂

**Comment by qsolutions-pl** [comment]
I still see it here:
https://github.com/magento/magento2/blob/2.4.7/composer.json

![image](https://github.com/magento/magento2/assets/4256191/90fb91d3-54c0-4aa3-b15e-9ac2afd2def9)

**Comment by Bashev** [comment]
@qsolutions-pl is expected, somehow Framework need to be loaded.

In the `magento-community-package` you can see only `Magento\\Setup`, because Framework comes from from `/vendor/magento/framework`

**Comment by engcom-Hotel** [comment]
Yes @hostep, the issue has been resolved in version 2.4.7.

Hence closing this issue.

Thanks
